src/poller_base.cpp

```diff
@@ -81,12 +82,13 @@ uint64_t zmq::poller_base_t::execute_timers ()
     //  Get the current time.
     const uint64_t current = _clock.now_ms ();
 
-    //   Execute the timers that are already due.
-    const timers_t::iterator begin = _timers.begin ();
-    const timers_t::iterator end = _timers.end ();
+    //  Execute the timers that are already due.
     uint64_t res = 0;
-    timers_t::iterator it = begin;
-    for (; it != end; ++it) {
+    timer_info_t timer_temp;
+    const timers_t::iterator end = _timers.end ();
+
+    for (timers_t::iterator it = _timers.begin (); it != end;
```

Comment:
The loop looks very weird now. Actually, there is no real loop variable anymore.

If reentrancy is really required, then I would suggest changing to a while loop: `while (!_timers.empty())`, and then just use `_timers.begin()` within the loop.

Another alternative might be to copy or swap `_timers` temporarily, but then handling `cancel_timer` gets more complicated.

Reply:
It doesn't stop when all timers are deleted, it stops earlier if there are no more due timers (`break`).

`for` loop is the right construct for this: it sets an initial value, condition to exit the loop and what happens in each iteration, all in one place. It doesn't strictly need to be an increment/decrement operator.

But if you feel that `while` is more suitable in this case I'll change it to `while`.

Reply:
Actually you're right, it can be `while (!_timers.empty())`. It's just a matter of preference.

Reply:
I agree that a `for` loop doesn't necessarily need to be restricted to decrement/increment in the iteration expression, but having an iteration expression that looks like the init expression is stretching that somewhat too far.

If you have strong feeling about keeping `for`, keep it, but I just think `while (!_timers.empty())` is much clearer here.

Reply:
Sorry for the slow replies, I'm pretty busy these days.

I think you're right about this.
That would eliminate the usage of `end ()` iterator and make things simpler.
I'll make changes for this.

Reply:
I've changed it to `while` now.
Actually I've changed it to `do ... while (!_timers.empty())` because at the beginning of `execute_timers ()` we already have this check once:
```
    //  Fast track.
    if (_timers.empty ())
        return 0;
```